buildengine/engine.go

```diff
@@ -149,6 +151,28 @@ func (e *Engine) Import(ctx context.Context, schema *schema.Module) {
 
 // Build attempts to build the specified modules, or all local modules if none are provided.
 func (e *Engine) Build(ctx context.Context, modules ...string) error {
+	return e.buildWithCallback(ctx, nil, modules...)
+}
+
+// Deploy attempts to build and deploy the specified modules, or all local modules if none are provided.
+func (e *Engine) Deploy(ctx context.Context, replicas int32, waitForDeployOnline bool, modules ...string) error {
+	if len(modules) == 0 {
+		modules = maps.Keys(e.modules)
+	}
+
+	expectedBuilds := make(map[string]struct{}, len(modules))
+	for _, name := range modules {
+		expectedBuilds[name] = struct{}{}
+	}
+
+	return e.buildWithCallback(ctx, func(ctx context.Context, module Module) error {
+		return Deploy(ctx, module, replicas, waitForDeployOnline, e.client)
```

Comment:
Having this wait for each deploy in turn will significantly slow down the deploy, as the build won't be able to continue until each group has fully deployed.

OTOH this will mean the deployed services will always have the correct dependencies up, so I'm a bit conflicted.

Probably okay for now, but we'll see how it goes. I think what would ideally happen is that deployments that don't have all the correct dependencies available wouldn't go live.

Reply:
Word! I was thinking, this is nice and reliable for now, but probably can be quicker in the future. I had it deploying without a wait, but def ran into issues with complex dependencies. Maybe something we can chat through to see if we can make this fast and reliable.